internal/storage/v2/memory/memory.go

```diff
@@ -0,0 +1,204 @@
+	"github.com/jaegertracing/jaeger/internal/tenancy"
+)
+
+const noServiceName = "OTLPResourceNoServiceName"
+
+// Store is an in-memory store of traces
+type Store struct {
+	sync.RWMutex
+	// Each tenant gets a copy of default config.
+	// In the future this can be extended to contain per-tenant configuration.
+	defaultConfig v1.Configuration
+	perTenant     map[string]*Tenant
+}
+
+// Tenant is an in-memory store of traces for a single tenant
+type Tenant struct {
+	sync.RWMutex
+	ids        []pcommon.TraceID
+	traces     map[pcommon.TraceID]ptrace.Traces
+	services   map[string]struct{}
+	operations map[string]map[v2api.Operation]struct{}
+	config     v1.Configuration
+	index      int
+}
+
+// NewStore creates an unbounded in-memory store
+func NewStore() *Store {
+	return WithConfiguration(v1.Configuration{MaxTraces: 0})
+}
+
+// WithConfiguration creates a new in memory storage based on the given configuration
+func WithConfiguration(cfg v1.Configuration) *Store {
+	return &Store{
+		defaultConfig: cfg,
+		perTenant:     make(map[string]*Tenant),
+	}
+}
+
+func newTenant(cfg v1.Configuration) *Tenant {
+	return &Tenant{
+		ids:        make([]pcommon.TraceID, cfg.MaxTraces),
+		traces:     map[pcommon.TraceID]ptrace.Traces{},
+		services:   map[string]struct{}{},
+		operations: map[string]map[v2api.Operation]struct{}{},
+		config:     cfg,
+	}
+}
+
+// getTenant returns the per-tenant storage.  Note that tenantID has already been checked for by the collector or query
+func (st *Store) getTenant(tenantID string) *Tenant {
+	st.RLock()
+	tenant, ok := st.perTenant[tenantID]
+	st.RUnlock()
+	if !ok {
+		st.Lock()
+		defer st.Unlock()
+		tenant, ok = st.perTenant[tenantID]
+		if !ok {
+			tenant = newTenant(st.defaultConfig)
+			st.perTenant[tenantID] = tenant
+		}
+	}
+	return tenant
+}
+
+// WriteTraces write the traces into the tenant by grouping all the spans with same trace id together.
+// The traces will not be saved as they are coming, rather they would be reshuffled, to understand this reshuffling
+// take an example of traces which have 2 resource spans, then these two resource spans have 2 scope spans each.
+// Every scope span consists of 2 spans with trace ids: 1 and 2. Now the final structure should look like:
+// For TraceID1: [ResourceSpan1:[ScopeSpan1:[Span(TraceID1)],ScopeSpan2:[Span(TraceID1)], ResourceSpan2:[ScopeSpan1:[Span(TraceID1)],ScopeSpan2:[Span(TraceID1)]
+// A similar structure will be there for TraceID2
+func (st *Store) WriteTraces(ctx context.Context, td ptrace.Traces) error {
+	m := st.getTenant(tenancy.GetTenant(ctx))
+	m.Lock()
+	defer m.Unlock()
+	sameTraceIDResourceSpans := make(map[pcommon.TraceID]ptrace.ResourceSpansSlice)
+	for i := 0; i < td.ResourceSpans().Len(); i++ {
+		sameTraceIDScopeSpans := make(map[pcommon.TraceID]ptrace.ScopeSpansSlice)
+		resourceSpan := td.ResourceSpans().At(i)
+		reshuffleScopeSpanAndSaveServiceAndOperation(m, resourceSpan, sameTraceIDScopeSpans)
```

Comment:
why does this have to combine reshuffle with "SaveServiceAndOperation"?

Reply:
reshuffling has nothing to do with updating the cache, you can separate those. Then loop through shuffled data again and apply changes to the cache.